pkgs/development/python-modules/pytorch/default.nix

```diff
@@ -30,23 +30,58 @@
   isPy3k, pythonOlder }:
 
 let
-  inherit (cudaPackages) cudatoolkit cudnn nccl;
+  inherit (cudaPackages) cudnn nccl;
 in
 
 # assert that everything needed for cuda is present and that the correct cuda versions are used
-assert !cudaSupport || (let majorIs = lib.versions.major cudatoolkit.version;
+assert cudaSupport -> (let majorIs = cudaPackages.cudaMajorVersion;
                         in majorIs == "9" || majorIs == "10" || majorIs == "11");
 
 # confirm that cudatoolkits are sync'd across dependencies
-assert !(MPISupport && cudaSupport) || mpi.cudatoolkit == cudatoolkit;
-assert !cudaSupport || magma.cudatoolkit == cudatoolkit;
+# TODO: verify it's OK to equality-compare cudaPackages (which are attrsets, scopes, or whatever)
+assert (MPISupport && cudaSupport) -> mpi.cudaPackages == cudaPackages;
+assert cudaSupport -> (magma.cudaPackages == cudaPackages);
 
 let
   setBool = v: if v then "1" else "0";
+
+  # cudatoolkit.cc is a passthru attribute
+  # that points at a compatible gcc version,
+  # we're not really using cudatoolkit;
+  # not sure we really need to hard-code gcc,
+  # but this is out of scope of current PR
+  # (which is: migrating pytorch to redist cuda)
+  inherit (cudaPackages.cudatoolkit) cc;
+
   cudatoolkit_joined = symlinkJoin {
-    name = "${cudatoolkit.name}-unsplit";
+    name = "cudatoolkit-root";
+
+    # The list of required cuda redist packages can be found e.g.
+    # at https://github.com/pytorch/pytorch/blob/b09769992f83f94150eaef2ab9d03c37b36da159/cmake/Summary.cmake#L85
+    # Not explicitly listed there are: nvml, nvtx, cccl
+    #
     # nccl is here purely for semantic grouping it could be moved to nativeBuildInputs
-    paths = [ cudatoolkit.out cudatoolkit.lib nccl.dev nccl.out ];
+    paths = with cudaPackages; [
+      cuda_nvcc
+      cuda_nvml_dev # <nvml.h>
+      cuda_nvtx # -llibNVToolsExt
+      cuda_cccl # <thrust/*>
+      cuda_nvprof # <cuda_profiler_api.h>
+      cuda_nvrtc
+      cuda_cudart
+      libcublas
+      libcufft
+      libcusolver
+      libcusparse
+      libcurand
+      nccl.dev
```

Comment:
it is preferable to use `lib.getDev` and `lib.getOut` over directly referring to outputs because this will keep functioning if outputs are being removed.